opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -2136,33 +2135,33 @@ namespace Opm
         // handle the out velcocity head
         const double area = segmentSet()[seg].crossArea();
         const EvalWell mass_rate = segment_mass_rates_[seg];
-        const EvalWell density = segment_densities_[seg];
+        const int seg_upwind = upwinding_segments_[seg];
+
+        // WARNING
+        // We disregard the derivatives from the upwind density to make sure derivatives
+        // wrt. to different segments dont get mixed.
+        const EvalWell density = segment_densities_[seg_upwind].value();
         const EvalWell out_velocity_head = mswellhelpers::velocityHead(area, mass_rate, density);
 
         resWell_[seg][SPres] -= out_velocity_head.value();
-        for (int pv_idx = 0; pv_idx < numWellEq; ++pv_idx) {
-            duneD_[seg][seg][SPres][pv_idx] -= out_velocity_head.derivative(pv_idx + numEq);
-        }
-
-        // calcuate the maximum cross-area among the segment and its inlet segments
-        double max_area = area;
-        for (const int inlet : segment_inlets_[seg]) {
-            const double inlet_area = segmentSet()[inlet].crossArea();
-            if (inlet_area > max_area) {
-                max_area = inlet_area;
-            }
-        }
+        duneD_[seg][seg][SPres][SPres] -= out_velocity_head.derivative(SPres + numEq);
+        duneD_[seg][seg][SPres][GTotal] -= out_velocity_head.derivative(GTotal + numEq);
+        duneD_[seg][seg_upwind][SPres][WFrac] -= out_velocity_head.derivative(WFrac + numEq);
+        duneD_[seg][seg_upwind][SPres][GFrac] -= out_velocity_head.derivative(GFrac + numEq);
 
         // handling the velocity head of intlet segments
         for (const int inlet : segment_inlets_[seg]) {
-            const EvalWell inlet_density = segment_densities_[inlet];
+            const int seg_upwind_inlet = upwinding_segments_[inlet];
+            const double inlet_area = segmentSet()[inlet].crossArea();
+            const EvalWell inlet_density = segment_densities_[seg_upwind_inlet].value();
```

Comment:
What is the reason we treat the `segment_densties_` and `segment_mass_rates_` differently? They are both associated with the flow compositions, which makes them both related to upwinding.

Reply:
segment_mass_rates_  contains the derivative wrt. the upwind fractions.

Reply:
I understand it now. It is really related to how to handle the upwinding for different parts.